myia/ir.py

```diff
@@ -1,7 +1,142 @@
-"""Intermediate representation definition."""
+"""Intermediate representation definition.
+
+Myia's main intermediate representation (IR) is a graph-based version of ANF.
+Each function definition (lambda) is defined as a graph, consisting of a series
+of function applications.
+
+A function can be applied to a node from another funtion's graph; this
+implicitly creates a nested function. Functions are first-class objects, so
+returning a nested function creates a closure.
+
+"""
+from enum import Enum, auto
+from typing import List, Union, Dict
+
+LiteralType = Union[str, int, float, 'PrimitiveOperation', 'Graph', None]
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
+        """Construct a graph."""
+        self.output = output
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are three types of nodes: Values, which are the result of function
+    applications; parameters; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a value, the first node input is the function to
+            apply, followed by the arguments.
+        value: The value of this node, if it is a constant.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a node."""
+        self.inputs: List[Node] = []
+        self.value: LiteralType = None
+        self.graph: Graph = None
+        self.debug: Dict = {}
+
+
+class Value(Node):
```

Comment:
I would prefer to call this class `Computation`, or `Result`.

Reply:
I find `Computation` a bit misleading, because no actual computation might be involved (e.g. higher order functions, identity, etc.) `Result` suggests to me that is the returned value of a function. `Value` on the other hand seems to be technically the [correct term](https://en.wikipedia.org/wiki/Value_(computer_science))?

Reply:
It would be a correct term for a node *in general*: a constant is also an r-value. A parameter is an l-value. Every node represents a value, which is why I think it's weird to use it specifically for function applications. How about `Application` or `Apply`?

Reply:
I like `Apply`.